refactor(gnovm): simplify DeepFill for constant value types

davd-gzl opened the pull request:
Fix: #4777
Fix: #4956

This revisits the DeepFill methods for the constant-only value types StringValue, BigintValue and BigdecValue raised in the audit.

After review, returning the value unchanged is the correct behavior rather than a sign of incorrect handling (see https://github.com/gnolang/gno/pull/4891#issuecomment-4333270764). DeepFill resolves RefValues to their concrete forms, so for a leaf value with no references the identity return is the base case of the recursion. These values also reach DeepFill legitimately in normal operation, for example when native bindings fill string parameters, so panicking or asserting that they are never called would be incorrect.

The change kept here is a small cleanup. ArrayValue, StructValue and HeapItemValue now fill their element TypedValues through the TypedValue.DeepFill wrapper instead of each inlining the same nil check and reassignment, so the nil guard lives in one place.

A regression test covers that DeepFill leaves the three constant types unchanged.

davd-gzl committed:
fix(gnovm): Concerns regarding the DeepFill method #4777

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: davd-gzl/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🔴 At least one of these user(s) reviewed the pull request: [aronpark1007 davd-gzl jefft0 notJoon omarsy MikaelVallenet] (with state "APPROVED")
    │       ├── 🟢 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

jefft0 commented:
Hi @Davphla . Can you fix the failed CI check for codecov?

davd-gzl committed:
test(gnolang): add unit tests for Constant DeepFill method in value types

thehowl reviewed (approved):
I'd say just make them panic, in correct code this adds no overhead anyway (because DeepFill is often called through an interface, not lending itself to optimization). So might as well always panic.

davd-gzl committed:
failing CI

davd-gzl committed:
Revert "failing CI"

This reverts commit 2b0ae1dd1037b298cc28c174fce40dd931bf5541.

davd-gzl committed:
fix(gnolang): prevent DeepFill on constant value types and update tests

davd-gzl committed:
refactor(gnolang): remove outdated comment

davd-gzl committed:
refactor: update incorrect comment changes

davd-gzl committed:
refactor: update incorrect comment changes

thehowl reviewed (approved):
LGTM, @ltzmaxwell for a second look

notJoon committed:
Merge branch 'master' into fix/deepfill

ltzmaxwell commented:
start looking.

ltzmaxwell commented:
Sorry for jumping in late and raising objections: 

A few thoughts on the design here, raising them for consideration. @davd-gzl @thehowl 
                                                                                                                                            
1. Returning self for constants is correct, not a contract violation.                                                                     
                                                                                                                                            
    DeepFill's contract is "resolve any RefValues in this value tree to their concrete forms." For a leaf value with no references —  StringValue, BigintValue, BigdecValue — return self is the identity case of the recursion. It's not `panic("shouldn't happen,")`,  it's the correct answer when called.                                                                                             
                                                                                                                                                                                                                                                                                                          
2. The wrapper shouldn't be a load-bearing choke point.                                                                                   
  
    Pre-PR, (*TypedValue).DeepFill was an optional ergonomic helper for the if tv.V != nil { tv.V = tv.V.DeepFill(store) } boilerplate. With the panics plus the new type-switch guard, it becomes `mandatory for correctness` — bypass it and you panic. That's a load-bearing convention with no compile-time enforcement. When MapValue.DeepFill (currently panic("not yet implemented")) gets a real implementation, the implementer must remember to route map element TypedValues through the wrapper, not call tv.V.DeepFill directly. Nothing in the type system tells them. That's a footgun the panics created.

3. The type-switch in the wrapper costs more than it saves.

    The guard exists only to skip a no-op (the leaf returning itself). What it costs:                                                         
      - Type-knowledge coupling: the wrapper has to enumerate which concrete types are leaves-without-refs. Add a future primitive type (of course low probability) and you have to update this list centrally rather than just implementing the interface correctly.                                                 
      - Layering inversion: a wrapper that depends on specific concrete implementations instead of the interface.                             
      - Reader friction: "why these three?" requires jumping to garbage_collector.go to discover the invariant.                                 
                                                                                                                                                                                             
Suggested alternative:
                                                                                                                                                                                                                                                  
  - Leaves return self for the three constant types (revert the panics). 
  - (*TypedValue).DeepFill reverts to:
      ```go                                                                   
      func (tv *TypedValue) DeepFill(store Store) {                                                                                             
          if tv.V != nil {                                                                                                                      
              tv.V = tv.V.DeepFill(store)
          }                                                                                                                                     
      }    
    ```                                                                                                                                   
  - Keep the in-package caller refactor (tv.DeepFill(store) instead of inlining the nil-check + reassign in ArrayValue/StructValue/HeapItemValue) — that's the genuine DRY win in this PR.

davd-gzl committed:
refactor(gnolang): revert DeepFill panics on constant types to return self

StringValue, BigintValue and BigdecValue are leaf constants with no
references, so returning self from DeepFill is the identity case of the
recursion, not a contract violation: native bindings routinely DeepFill
string/bigint return values. Drop the panics and the TypedValue.DeepFill
type-switch guard that only existed to skip them, keeping the in-package
caller refactor (ArrayValue/StructValue/HeapItemValue use tv.DeepFill).

davd-gzl committed:
test(gnolang): simplify TestConstantValuesDeepFill to a table loop

Collapse the three near-identical subtests into a single loop over the
constant value types, asserting DeepFill returns each one unchanged.

davd-gzl committed:
Merge remote-tracking branch 'fork' into review-fixes/4891

davd-gzl commented:
Agreed, thank you @ltzmaxwell, I fixed and simplified this work: https://github.com/gnolang/gno/pull/4891/commits/6f742bc42da99801e1cd62fc2be6117105ad0186
Now this PR is only cosmetic + added regression test

davd-gzl committed:
Merge branch 'master' into fix/deepfill
